Assets/Simon/Scripts/Sequence.cs

```diff
@@ -0,0 +1,55 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using System;
+
+public class Sequence : MonoBehaviour
+{
+    public List<Colors> ColorSequence { get; private set; }
+
+    private int sequenceIndex = 0;
+    public void CreateColorSequence(int sequenceLength)
+    {
+        if(sequenceLength <= 0)
+        {
+            Debug.LogError("The length of the sequence should be larger than 0");
+        }
+
+        ColorSequence = new List<Colors>();
+
+        for(int i = 0; i < sequenceLength; i++)
+        {
+            ExtendColorSequence();
+        }
+    }
+
+    public void IterateSequence()
+    {
+       sequenceIndex++;
+    }
+
+    public bool SequenceReachedLastItem()
+    {
+        return sequenceIndex >= ColorSequence.Count;
+    }
+
+    public void ResetSequenceIndex()
+    {
+        sequenceIndex = 0;
+    }
+
+    private Colors getRandomColor()
+    {
+        return (Colors)UnityEngine.Random.Range(0, 4);

```

Comment:
If you change the indexes, or add more members (you have the max value hard-coded) to the enum, this will be invalid. 

I would do: 
```csharp        
        var allColors = Enum.GetValues(typeof(Colors));
        var randomColors = (Colors)allColors.GetValue(UnityEngine.Random.Range(0, allColors.Length));
```
Then it's always the correct range.

I can't remember if Unity random range is inclusive or exclusive, so it might have to be ```colors.Length - 1```

Reply:
Yes, I kept it very straightforward because the game is based on exactly 4 colors. But you are right, I'll do that.
By the way I am a bit confused about 
`var allColors = Enum.GetValues(typeof(Colors));`

What is it's return type. Because I tried to do the same thing without using var, but couldn't do it. 
I don't know much about var actually.

Reply:
Yeah i just used var for an example, it is still strongly typed, but the compiler works out what type it is from the expression that is evaluated on that line. It is called 'implicit type' or 'inferred type'. It's just for local variables, you can't use it for a class member.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/implicitly-typed-local-variables

The return type would be ```Array```, which is the base class for all arrays in c#, so you have to cast it to the type of your enum. 

You could also cast the array rather than the element, which might be better than my example:

```csharp
Colors[] allColors = (Colors[])Enum.GetValues(typeof(Colors));
```
The Enum type has some static methods like this so you can get the values or the names, which is nice.

Reply:
I see. Actually I moved allColors = (Colors[])Enum.GetValues(typeof(Colors)) to a private field, since we only need get all the colors in the array only once. We don't need to call Enum.GetValues() every time we want to pick a random color.

`private Colors[] allColors = (Colors[])Enum.GetValues(typeof(Colors));`

Reply:
Yeah looks good - doing it on one line is fine. I just did it on two for the sake of example.